### PR TITLE
arch: arm64: mmu: avoid using of set/way cache instructions

### PR DESCRIPTION
**Pull request by xakep-amatop**
Exposing set/way cache maintenance to a virtual machine is unsafe, not least because the instructions are not permission-checked, but also because they are not broadcast between CPUs.

Unfortunately, set/way ops are not easily virtualized by Xen. S/W emulation is disabled, because IP-MMU is active for Dom0. IP-MMU is a IO-MMU made by Renesas, as any good IO-MMU, it shares page-tables with CPU. Trying to emulate S/W with IP-MMU active will lead to IO-MMU faults. So if we build Zephyr as a Xen Initial domain, it won't work with cache management support enabled.

In this commit, VA data invalidate of mapped pages is used instead of using set/way instructions and all set/way cache operation have been disabled for Xen domains.

Related pull for xen-lib: https://github.com/xen-troops/zephyr-xenlib/pull/46.

### CODE REVIEW

File: arch/arm64/core/mmu.c [hunk 1]
```diff
@@ -629,7 +630,17 @@ static int __add_map(struct arm_mmu_ptables *ptables, const char *name,
 	__ASSERT(((virt | phys | size) & (CONFIG_MMU_PAGE_SIZE - 1)) == 0,
 		 "address/size are not page aligned\n");
 	desc |= phys;
-	return set_mapping(ptables, virt, size, desc, may_overwrite);
+	ret = set_mapping(ptables, virt, size, desc, may_overwrite);
+#ifdef CONFIG_DT_HAS_XEN_XEN_ENABLED
+	if (!ret) {
+		unsigned int mem_type = MT_TYPE(attrs);
+
+		if ((mem_type == MT_NORMAL) || (mem_type == MT_NORMAL_NC)) {
```
Comment (lorc) [inline arch/arm64/core/mmu.c]: `NC` stands non-cacheable, right? Should it be invalidated? I am more interested in `WB` mappings

File: arch/arm64/core/cache.c [hunk 2]
```diff
@@ -139,6 +139,9 @@ int arm64_dcache_range(void *addr, size_t size, int op)
  */
 int arm64_dcache_all(int op)
 {
+#ifdef CONFIG_DT_HAS_XEN_XEN_ENABLED
```
Comment (lorc) [inline arch/arm64/core/cache.c]: You should revert the workaround patch. In a separate commit.

File: arch/arm64/core/mmu.c [hunk 3]
```diff
@@ -629,7 +630,17 @@ static int __add_map(struct arm_mmu_ptables *ptables, const char *name,
 	__ASSERT(((virt | phys | size) & (CONFIG_MMU_PAGE_SIZE - 1)) == 0,
 		 "address/size are not page aligned\n");
 	desc |= phys;
-	return set_mapping(ptables, virt, size, desc, may_overwrite);
+	ret = set_mapping(ptables, virt, size, desc, may_overwrite);
+#ifdef CONFIG_DT_HAS_XEN_XEN_ENABLED
```
Comment (lorc) [inline arch/arm64/core/mmu.c]: Do you really need this `ifdef`? It is better to make this variant only one option. No need to support two different approaches.

As you pointed out, set/way  does not support L3 cache, so your approach will be beneficial for all users, not only for Xen domains.

**Review by lorc — COMMENTED** [review]
I like the change. But ideally, we need 3 commits:

1. Revert previous workaround
2. Implement your fix, but without `ifdef`
3. Remove `sys_cache_data_invd_all` function at all.

Second two commits may be upstreamed right away

File: arch/arm64/core/mmu.c (hunk 1 above)
Comment (xakep-amatop) [reply arch/arm64/core/mmu.c]: I deleted these checks, pls look at commit message for more details

File: arch/arm64/core/cache.c (hunk 2 above)
Comment (xakep-amatop) [reply arch/arm64/core/cache.c]: done

File: arch/arm64/core/mmu.c (hunk 3 above)
Comment (xakep-amatop) [reply arch/arm64/core/mmu.c]: done

**Comment by xakep-amatop** [comment]
> I like the change. But ideally, we need 3 commits:
> 
>     1. Revert previous workaround
> 
>     2. Implement your fix, but without `ifdef`
> 
>     3. Remove `sys_cache_data_invd_all` function at all.
> 
> 
> Second two commits may be upstreamed right away

done

File: drivers/xen/regions.c [hunk 4]
```diff
@@ -133,6 +133,31 @@ static int addr_is_in_buffer(uint8_t *ptr, sys_mem_blocks_t *block)
...
+	}
+
+	return false;
+}
+
+void xen_region_map(void *ptr, size_t size)
+{
+	if (addr_from_extended_region(ptr)) {
+		arch_mem_map(ptr, (uintptr_t)ptr, size, K_MEM_CACHE_WB | K_MEM_PERM_RW);
+	}
```
Comment (lorc) [inline drivers/xen/regions.c:154]: It will be nice to output some error/warning message otherwise

File: drivers/xen/regions.c [hunk 5]
```diff
@@ -133,6 +133,31 @@ static int addr_is_in_buffer(uint8_t *ptr, sys_mem_blocks_t *block)
...
+}
+
+void xen_region_unmap(void *ptr, size_t size)
+{
+	if (addr_from_extended_region(ptr)) {
+		arch_mem_unmap(ptr, size);
+	}
```
Comment (lorc) [inline drivers/xen/regions.c:163]: It will be nice to output some error/warning message otherwise

**Review by lorc — APPROVED** [review]
Awesome, thanks!

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuck@epam.com>`

File: include/zephyr/xen/regions.h [hunk 6]
```diff
@@ -21,4 +21,18 @@ void *xen_region_get_pages(size_t nr_pages);
  */
 void xen_region_put_pages(void *ptr, size_t nr_pages);
 
+/**
+ * @brief Map extended region memory
+ * @param ptr - pointer to the pages, allocated by xen_region_get_pages call
+ * @param size - allocated memory size in bytes
```
Comment (lorc) [inline include/zephyr/xen/regions.h]: Could we use number of pages there? It will be most logical

**Review by lorc — COMMENTED** [review]
Sorry for the late comment, but I just spotted this

**Comment by xakep-amatop** [comment]
@lorc addressed all your comments and add tag to commits

**Review by firscity — APPROVED** [review]
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

**Comment by xakep-amatop** [comment]
added ack tag to commits
note: changes aren't tetsed yet with aos app

File: arch/arm64/core/mmu.c [hunk 7]
```diff
@@ -629,7 +630,15 @@ static int __add_map(struct arm_mmu_ptables *ptables, const char *name,
 	__ASSERT(((virt | phys | size) & (CONFIG_MMU_PAGE_SIZE - 1)) == 0,
 		 "address/size are not page aligned\n");
 	desc |= phys;
-	return set_mapping(ptables, virt, size, desc, may_overwrite);
+	ret = set_mapping(ptables, virt, size, desc, may_overwrite);
+	if (ret) {
+		return ret;
+	}
+
+	if (MT_TYPE(attrs) == MT_NORMAL || MT_TYPE(attrs) == MT_NORMAL_WT) {
```
Comment (xakep-amatop) [inline arch/arm64/core/mmu.c]: only add this condition + change description of commit message related to the below line

**Comment by xakep-amatop** [comment]
Added changes only to the next commit, cd4f6612bd8d2601094c3e43fa1e370a8ce9079b: perform data cache invalidation after TLB invalidation.

**Comment by xakep-amatop** [comment]
changes tested and work without crashes

**Review by lorc — APPROVED** [review]
I am still fine with the changes, so you can leave the tags.

I'm bit concerned that `invalidate_tlb_all` now is called inside a loop, but frankly I am not expecting any performance hit.

**Comment by xakep-amatop** [comment]
> I'm bit concerned that `invalidate_tlb_all` now is called inside a loop, but frankly I am not expecting any performance hit.

I can create two additional loops after tlb ivalidate just for dcache invalidating if you want

**Comment by lorc** [comment]
> I can create two additional cycles after tlb ivalidate just for dcache invalidating if you want

"Loops", you wanted say "loops" :) In my opinion it is better to have a separate loop that invalidates all the mapped regions. This will produce a cleaner diff and less WTF moments as a side effect. But it is up to you, really.

**Comment by xakep-amatop** [comment]
> > I can create two additional cycles after tlb ivalidate just for dcache invalidating if you want
> 
> "Loops", you wanted say "loops" :) In my opinion it is better to have a separate loop that invalidates all the mapped regions. This will produce a cleaner diff and less WTF moments as a side effect. But it is up to you, really.

updated, pls, review

**Comment by lorc** [comment]
Yep, I'm fine with this. I believe it can be merged if there are no other dependencies

**Pull request merged**